File: agents/node/src/health.ts
import type { Framework } from './detect'
import type { RukkieConfig } from './setup'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyApp = any

export function registerHealthRoute(app: AnyApp, framework: Framework, config: RukkieConfig): void {
  if (framework === 'express') {
    app.get('/__rukkie/health', async (_req: AnyApp, res: AnyApp) => {
      const body = await buildHealthResponse(config)
      res.status(body.status === 'ok' ? 200 : 207).json(body)
    })
  } else if (framework === 'fastify') {
    app.get('/__rukkie/health', async (_request: AnyApp, reply: AnyApp) => {
      const body = await buildHealthResponse(config)
      reply.status(body.status === 'ok' ? 200 : 207).send(body)
    })
  }
}

async function buildHealthResponse(config: RukkieConfig): Promise<HealthResponse> {
  const deps: Record<string, string> = {}

  if (config.dependencies) {
    await Promise.all(
      Object.entries(config.dependencies).map(async ([name, check]) => {
        try {
          const ok = await check()
          deps[name] = ok ? 'connected' : 'error'
        } catch {
          deps[name] = 'error'
        }
      })
    )
  }

  const allOk = Object.values(deps).every((v) => v === 'connected')
  const status = Object.keys(deps).length === 0 ? 'ok' : allOk ? 'ok' : 'degraded'

  return {
    status,
    service: config.serviceName,
    dependencies: deps,
  }
}

interface HealthResponse {
  status: 'ok' | 'degraded'
  service: string
  dependencies: Record<string, string>
}

File: agents/python/rukkie_agent/health.py
from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING, Any, Dict

if TYPE_CHECKING:
    from .setup import RukkieConfig


def register_health_endpoint(app: Any, framework: str, config: "RukkieConfig") -> None:
    if framework == "fastapi":
        _register_fastapi(app, config)
    elif framework == "flask":
        _register_flask(app, config)
    elif framework == "starlette":
        _register_fastapi(app, config)  # same API


def _register_fastapi(app: Any, config: "RukkieConfig") -> None:
    async def health_handler():  # type: ignore[misc]
        return await _build_response(config)

    app.add_api_route("/__rukkie/health", health_handler, methods=["GET"])


def _register_flask(app: Any, config: "RukkieConfig") -> None:
    from flask import jsonify

    @app.route("/__rukkie/health", methods=["GET"])
    def rukkie_health():  # type: ignore[misc]
        body = asyncio.run(_build_response(config))
        status = 200 if body["status"] == "ok" else 207
        return jsonify(body), status


async def _build_response(config: "RukkieConfig") -> Dict[str, Any]:
    deps: Dict[str, str] = {}

    if config.dependencies:
        async def check_one(name: str, fn: Any) -> None:
            try:
                result = fn()
                if asyncio.iscoroutine(result):
                    result = await result
                deps[name] = "connected" if result else "error"
            except Exception:
                deps[name] = "error"

        await asyncio.gather(*(check_one(n, f) for n, f in config.dependencies.items()))

    all_ok = all(v == "connected" for v in deps.values())
    status = "ok" if (not deps or all_ok) else "degraded"

    return {
        "status": status,
        "service": config.service_name,
        "dependencies": deps,
    }
